### PR TITLE
Add ability to write a filtered blob through a stream

### PR DESCRIPTION
**Pull request by carlosmn**
In the fashion of `_fromchunks()` but leaving the caller with the decision of when to add data instead of making them jump through hoops to reverse the control inversion.

### CODE REVIEW

File: include/git2/blob.h [hunk 1]
```diff
@@ -192,6 +192,49 @@ GIT_EXTERN(int) git_blob_create_fromchunks(
 	void *payload);
 
 /**
+ * Create  stream to write a new blob into the object db
```
Comment (nulltoken) [inline include/git2/blob.h]: Two spaces here after `Create`

Comment (carlosmn) [inline include/git2/blob.h]: The 'a' is invisible :p

File: src/blob.c [hunk 2]
```diff
@@ -334,6 +334,99 @@ int git_blob_create_fromchunks(
...
+	int error;
+	git_buf path = GIT_BUF_INIT;
+	blob_writestream *stream;
+
+	assert(out && repo);
+
+	stream = git__calloc(1, sizeof(blob_writestream));
+	GITERR_CHECK_ALLOC(stream);
+
+	stream->repo     = repo;
```
Comment (ethomson) [inline src/blob.c]: so many spaces!

File: src/blob.c [hunk 3]
```diff
@@ -334,6 +334,99 @@ int git_blob_create_fromchunks(
...
+	error = git_blob__create_from_paths(out, NULL, stream->repo, stream->fbuf.path_lock,
+					    stream->hintpath, 0, !!stream->hintpath);
+
+cleanup:
+	blob_writestream_free(_stream);
+	return error;
+
+}
+
+
```
Comment (Therzok) [inline src/blob.c]: Zomg, 2 newlines :trollface:

File: include/git2/blob.h [hunk 4]
```diff
@@ -192,6 +192,49 @@ GIT_EXTERN(int) git_blob_create_fromchunks(
...
+/**
+ * Close the stream and write the blob to the object db
+ *
+ * The stream will be closed and freed.
+ *
+ * @param out the id of the new blob
+ * @param stream the stream to close
+ * @return 0 or an error code
+ */
+GIT_EXTERN(int) git_blob_create_fromstream_end(
```
Comment (carlosmn) [inline include/git2/blob.h]: Rename this to `_commit()` so we match what we do for the indexer and filebuf.

**Comment by ethomson** [comment]
Do we need the `git_blob_create_fromstream_commit` function?  It's been a little while since I played with the streams, but I feel like this is what `writestream->close` was intended to accomplish.  It would be nice to keep the committing / aborting encapsulated to the writestream itself, and not a separate function just for the blob streaming creation.

**Comment by carlosmn** [comment]
We need a function which will take a steam, close it and report back the object id. The closing works as committing the filebuf via the interface just fine, but this interface doesn't have a function that will let us report back on the resulting object id.

We need *some* extra function so we know what the object id is. Alternatively we could make it a `git_odb_stream` instead of a `git_writestream`. It does have a `_finalize_write()` function which tells you what the object id ends up being.

**Comment by ethomson** [comment]
> Alternatively we could make it a git_odb_stream instead of a git_writestream. It does have a _finalize_write() function which tells you what the object id ends up being.

That sounds quite nice to me.

**Comment by carlosmn** [comment]
Using a `git_odb_stream` won't work here. That type represents the final stream of data to put into the object db and it performs the hashing in the "methods" instead of leaving it up to the implementer.

So if we want to stay away from `git_blob_create_fromstream_commit`  we'd need to come up with a different interface for this kind of stream... or maybe add a new optional method on `git_writestream` that's something like `git_writestream_close_hashed()` which reports back the id.

**Comment by carlosmn** [comment]
Since we haven't found a better name so let's :ship: this so we get the feature and we can hide this in the other languages behind their own stream implementations.

**Pull request merged**